File: src/server/services/fileService.ts
import crypto from 'node:crypto'
import path from 'node:path'
import fs from 'fs-extra'
import { glob } from 'glob'

export interface FileMetadata {
  path: string
  name: string
  relativePath: string
  directory: string
  size: number
  lastModified: Date
  created: Date
  hash: string
  lineCount?: number
  wordCount?: number
}

export interface DirectoryInfo {
  path: string
  name: string
  fileCount: number
}

export interface FileContent {
  path: string
  name: string
  content: string
  metadata: FileMetadata
  frontmatter?: Record<string, unknown>
}

export class FileService {
  private watchDirectory: string
  private manifestPath: string

  constructor(watchDirectory: string = './docs') {
    this.watchDirectory = path.resolve(watchDirectory)
    this.manifestPath = path.join(this.watchDirectory, '.specmgr-manifest.json')
  }

  /**
   * Get all Markdown files in the watched directory
   */
  async getFiles(
    options: {
      pathFilter?: string
      recursive?: boolean
      sortBy?: 'name' | 'modified' | 'size'
      order?: 'asc' | 'desc'
    } = {}
  ): Promise<{ files: FileMetadata[]; directories: DirectoryInfo[] }> {
    const { pathFilter = '', recursive = true, sortBy = 'name', order = 'asc' } = options

    try {
      // Ensure watch directory exists
      await fs.ensureDir(this.watchDirectory)

      // Build glob pattern
      const pattern = recursive
        ? path.join(this.watchDirectory, pathFilter, '**/*.md')
        : path.join(this.watchDirectory, pathFilter, '*.md')

      // Find all markdown files
      const filePaths = await glob(pattern, { ignore: ['**/node_modules/**', '**/.git/**'] })

      // Get file metadata
      const files: FileMetadata[] = []
      const directories = new Set<string>()

      for (const filePath of filePaths) {
        try {
          const stats = await fs.stat(filePath)
          const relativePath = path.relative(this.watchDirectory, filePath)
          const directory = path.dirname(relativePath)

          directories.add(directory)

          const content = await fs.readFile(filePath, 'utf-8')
          const hash = crypto.createHash('sha1').update(content).digest('hex')

          const metadata: FileMetadata = {
            path: filePath,
            name: path.basename(filePath),
            relativePath,
            directory: directory === '.' ? '' : directory,
            size: stats.size,
            lastModified: stats.mtime,
            created: stats.birthtime,
            hash,
            lineCount: content.split('\n').length,
            wordCount: content.split(/\s+/).filter(word => word.length > 0).length,
          }

          files.push(metadata)
        } catch (error) {
          console.warn(`Failed to process file ${filePath}:`, error)
        }
      }

      // Sort files
      files.sort((a, b) => {
        let comparison = 0
        switch (sortBy) {
          case 'name':
            comparison = a.name.localeCompare(b.name)
            break
          case 'modified':
            comparison = a.lastModified.getTime() - b.lastModified.getTime()
            break
          case 'size':
            comparison = a.size - b.size
            break
        }
        return order === 'desc' ? -comparison : comparison
      })

      // Get directory info
      const directoryInfos: DirectoryInfo[] = []
      for (const dir of directories) {
        if (dir) {
          const dirPath = path.join(this.watchDirectory, dir)
          const dirFiles = files.filter(f => f.directory === dir)
          directoryInfos.push({
            path: dirPath,
            name: path.basename(dir),
            fileCount: dirFiles.length,
          })
        }
      }

      return { files, directories: directoryInfos }
    } catch (error) {
      console.error('Error getting files:', error)
      throw new Error('Failed to retrieve files')
    }
  }

  /**
   * Get content and metadata for a specific file
   */
  async getFileContent(relativePath: string): Promise<FileContent> {
    try {
      const filePath = path.join(this.watchDirectory, relativePath)

      // Security check - ensure file is within watch directory
      const resolvedPath = path.resolve(filePath)
      const resolvedWatchDir = path.resolve(this.watchDirectory)
      if (!resolvedPath.startsWith(resolvedWatchDir)) {
        throw new Error('Access denied: File is outside watch directory')
      }

      // Check if file exists
      if (!(await fs.pathExists(filePath))) {
        throw new Error('File not found')
      }

      // Read file content
      const content = await fs.readFile(filePath, 'utf-8')
      const stats = await fs.stat(filePath)

      // Generate hash
      const hash = crypto.createHash('sha1').update(content).digest('hex')

      // Parse frontmatter (basic implementation)
      const frontmatter = this.parseFrontmatter(content)
      const contentWithoutFrontmatter = this.removeFrontmatter(content)

      const metadata: FileMetadata = {
        path: filePath,
        name: path.basename(filePath),
        relativePath,
        directory: path.dirname(relativePath),
        size: stats.size,
        lastModified: stats.mtime,
        created: stats.birthtime,
        hash,
        lineCount: content.split('\n').length,
        wordCount: contentWithoutFrontmatter.split(/\s+/).filter(word => word.length > 0).length,
      }

      return {
        path: filePath,
        name: path.basename(filePath),
        content: contentWithoutFrontmatter,
        metadata,
        frontmatter,
      }
    } catch (error) {
      console.error(`Error reading file ${relativePath}:`, error)
      throw error
    }
  }

  /**
   * Get the current manifest (file hash tracking)
   */
  async getManifest(): Promise<Record<string, string>> {
    try {
      if (await fs.pathExists(this.manifestPath)) {
        return await fs.readJson(this.manifestPath)
      }
      return {}
    } catch (error) {
      console.warn('Error reading manifest:', error)
      return {}
    }
  }

  /**
   * Update the manifest with current file hashes
   */
  async updateManifest(fileHashes: Record<string, string>): Promise<void> {
    try {
      await fs.writeJson(this.manifestPath, fileHashes, { spaces: 2 })
    } catch (error) {
      console.error('Error updating manifest:', error)
      throw error
    }
  }

  /**
   * Get files that have changed since last manifest update
   */
  async getChangedFiles(): Promise<{
    added: string[]
    modified: string[]
    deleted: string[]
  }> {
    try {
      const currentManifest = await this.getManifest()
      const { files } = await this.getFiles()

      const currentFiles = new Map<string, string>()
      files.forEach(file => {
        currentFiles.set(file.relativePath, file.hash)
      })

      const added: string[] = []
      const modified: string[] = []
      const deleted: string[] = []

      // Check for added and modified files
      for (const [filePath, hash] of currentFiles) {
        if (!currentManifest[filePath]) {
          added.push(filePath)
        } else if (currentManifest[filePath] !== hash) {
          modified.push(filePath)
        }
      }

      // Check for deleted files
      for (const filePath of Object.keys(currentManifest)) {
        if (!currentFiles.has(filePath)) {
          deleted.push(filePath)
        }
      }

      return { added, modified, deleted }
    } catch (error) {
      console.error('Error getting changed files:', error)
      throw error
    }
  }

  /**
   * Parse frontmatter from markdown content
   */
  private parseFrontmatter(content: string): Record<string, unknown> | undefined {
    const frontmatterRegex = /^---\n([\s\S]*?)\n---\n/
    const match = content.match(frontmatterRegex)

    if (!match) return undefined

    try {
      const yamlContent = match[1]
      const frontmatter: Record<string, unknown> = {}

      // Basic YAML parsing (for simple key-value pairs)
      const lines = yamlContent.split('\n')
      for (const line of lines) {
        const colonIndex = line.indexOf(':')
        if (colonIndex > 0) {
          const key = line.substring(0, colonIndex).trim()
          const value = line.substring(colonIndex + 1).trim()

          // Remove quotes if present
          const cleanValue = value.replace(/^["']|["']$/g, '')

          // Try to parse as array if it looks like one
          if (cleanValue.startsWith('[') && cleanValue.endsWith(']')) {
            try {
              frontmatter[key] = JSON.parse(cleanValue)
            } catch {
              frontmatter[key] = cleanValue
            }
          } else {
            frontmatter[key] = cleanValue
          }
        }
      }

      return Object.keys(frontmatter).length > 0 ? frontmatter : undefined
    } catch (error) {
      console.warn('Error parsing frontmatter:', error)
      return undefined
    }
  }

  /**
   * Remove frontmatter from markdown content
   */
  private removeFrontmatter(content: string): string {
    const frontmatterRegex = /^---\n([\s\S]*?)\n---\n/
    return content.replace(frontmatterRegex, '')
  }
}

File: src/server/app/services/file_service.py
"""ファイルサービス."""

import hashlib
from datetime import datetime
from pathlib import Path

from app.core.config import settings
from app.models.api_models import (
    DirectoryInfo,
    FileContent,
    FileMetadata,
    FilesResponse,
)


class FileService:
    """ファイル管理サービス."""

    def __init__(self) -> None:
        self.docs_path = Path(settings.documents_path).resolve()

    async def get_files(
        self,
        path: str | None = None,
        recursive: bool = True,
        sort_by: str = "name",
        order: str = "asc",
    ) -> FilesResponse:
        """
        ファイル一覧取得.

        Args:
            path: 対象パス
            recursive: 再帰的検索
            sort_by: ソート基準
            order: ソート順序

        Returns:
            ファイル一覧
        """
        target_path = self.docs_path
        if path:
            target_path = self.docs_path / path

        if not target_path.exists():
            return FilesResponse(files=[], directories=[], totalCount=0)

        files = []
        directories = []

        # ファイルとディレクトリを取得
        if recursive:
            for file_path in target_path.rglob("*.md"):
                if file_path.is_file():
                    metadata = await self._get_file_metadata(file_path)
                    files.append(metadata)
        else:
            for item in target_path.iterdir():
                if item.is_file() and item.suffix.lower() in [".md", ".markdown"]:
                    metadata = await self._get_file_metadata(item)
                    files.append(metadata)
                elif item.is_dir():
                    dir_info = await self._get_directory_info(item)
                    directories.append(dir_info)

        # ソート
        reverse = order == "desc"
        if sort_by == "name":
            files.sort(key=lambda x: x.name, reverse=reverse)
        elif sort_by == "modified":
            files.sort(key=lambda x: x.last_modified, reverse=reverse)
        elif sort_by == "size":
            files.sort(key=lambda x: x.size, reverse=reverse)

        return FilesResponse(
            files=files,
            directories=directories,
            totalCount=len(files) + len(directories),
        )

    async def get_file_content(self, file_path: str) -> FileContent:
        """
        ファイル内容取得.

        Args:
            file_path: ファイルパス

        Returns:
            ファイル内容
        """
        full_path = self.docs_path / file_path

        if not full_path.exists():
            raise FileNotFoundError(f"ファイルが見つかりません: {file_path}")

        if not full_path.is_file():
            raise ValueError(f"指定されたパスはファイルではありません: {file_path}")

        # ファイル内容を読み込み
        try:
            with open(full_path, encoding="utf-8") as f:
                content = f.read()
        except UnicodeDecodeError:
            # UTF-8で読めない場合はShift_JISで試す
            with open(full_path, encoding="shift_jis") as f:
                content = f.read()

        # メタデータを取得
        metadata = await self._get_file_metadata(full_path)

        # フロントマター解析（簡易版）
        frontmatter = None
        if content.startswith("---\n"):
            try:
                end_pos = content.find("\n---\n", 4)
                if end_pos > 0:
                    frontmatter_text = content[4:end_pos]
                    # 簡易的なYAMLパース（実際にはyamlライブラリを使用すべき）
                    frontmatter = {}
                    for line in frontmatter_text.split("\n"):
                        if ":" in line:
                            key, value = line.split(":", 1)
                            frontmatter[key.strip()] = value.strip()
            except Exception:  # noqa: S110
                # フロントマター解析に失敗してもエラーにしない
                pass

        return FileContent(
            path=file_path,
            name=full_path.name,
            content=content,
            metadata=metadata,
            frontmatter=frontmatter,
        )

    async def _get_file_metadata(self, file_path: Path) -> FileMetadata:
        """ファイルメタデータを取得."""
        stat = file_path.stat()
        relative_path = str(file_path.relative_to(self.docs_path))

        # ファイルハッシュを計算
        file_hash = await self._calculate_file_hash(file_path)

        # 行数とワード数を計算
        line_count, word_count = await self._count_lines_words(file_path)

        return FileMetadata(
            name=file_path.name,
            path=str(file_path),
            relativePath=relative_path,
            directory=str(file_path.parent.relative_to(self.docs_path)),
            size=stat.st_size,
            lastModified=datetime.fromtimestamp(stat.st_mtime),
            created=datetime.fromtimestamp(stat.st_ctime),
            hash=file_hash,
            lineCount=line_count,
            wordCount=word_count,
        )

    async def _get_directory_info(self, dir_path: Path) -> DirectoryInfo:
        """ディレクトリ情報を取得."""
        relative_path = str(dir_path.relative_to(self.docs_path))

        # ディレクトリ内のMarkdownファイル数をカウント
        file_count = len(list(dir_path.rglob("*.md")))

        return DirectoryInfo(
            name=dir_path.name,
            path=str(dir_path),
            relativePath=relative_path,
            fileCount=file_count,
        )

    async def _calculate_file_hash(self, file_path: Path) -> str:
        """ファイルのSHA-1ハッシュを計算."""
        hasher = hashlib.sha1()  # noqa: S324
        try:
            with open(file_path, "rb") as f:
                for chunk in iter(lambda: f.read(4096), b""):
                    hasher.update(chunk)
            return hasher.hexdigest()
        except Exception:
            return ""

    async def _count_lines_words(self, file_path: Path) -> tuple[int, int]:
        """ファイルの行数とワード数をカウント."""
        try:
            with open(file_path, encoding="utf-8") as f:
                content = f.read()

            line_count = len(content.splitlines())
            word_count = len(content.split())

            return line_count, word_count
        except Exception:
            return 0, 0
